numpy/core/numeric.py

```diff
@@ -2341,6 +2361,11 @@ def within_tol(x, y, atol, rtol):
     x = asanyarray(a)
     y = asanyarray(b)
 
+    if not np.can_cast(x.dtype, y.dtype, casting=casting):
+        raise TypeError("Cannot compare types '%s' and '%s' with casting rule"
+                        " %s" % (str(x.dtype), str(y.dtype), casting))
```

Comment:
```suggestion
        raise TypeError(f"Cannot compare types {x.dtype!r} and {y.dtype!r} "
                        f"with casting rule {casting!r}")
```
NIT: Style suggestion, _i.e._ f-strings are awesome and their usage should be encouraged.